File: app/src/main/java/org/torproject/android/ui/AppManagerActivity.java
/* Copyright (c) 2009, Nathan Freitas, Orbot / The Guardian Project - http://openideals.com/guardian */
/* See LICENSE for licensing information */

package org.torproject.android.ui;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.GridView;
import android.widget.ImageView;
import android.widget.ListAdapter;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import org.torproject.android.BuildConfig;
import org.torproject.android.R;
import org.torproject.android.service.OrbotConstants;
import org.torproject.android.service.util.Prefs;
import org.torproject.android.service.vpn.TorifiedApp;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;
import java.util.StringTokenizer;

public class AppManagerActivity extends AppCompatActivity implements OnClickListener, OrbotConstants {


    class TorifiedAppWrapper {
        String header = null;
        String subheader = null;
        TorifiedApp app = null;
    }

    private PackageManager pMgr = null;
    private SharedPreferences mPrefs = null;
    private GridView listAppsAll;
    private ListAdapter adapterAppsAll;
    private ProgressBar progressBar;
    private ArrayList<String> alSuggested;

    /**
     * @return true if the app is "enabled", not Orbot, and not in
     * {@link #BYPASS_VPN_PACKAGES}
     */
    public static boolean includeAppInUi(ApplicationInfo applicationInfo) {
        if (!applicationInfo.enabled) return false;
        if (BYPASS_VPN_PACKAGES.contains(applicationInfo.packageName)) return false;
        return !BuildConfig.APPLICATION_ID.equals(applicationInfo.packageName);
    }

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        pMgr = getPackageManager();

        this.setContentView(R.layout.layout_apps);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        listAppsAll = findViewById(R.id.applistview);
        progressBar = findViewById(R.id.progressBar);

        //need a better way to manage this list
        alSuggested = new ArrayList<>();
        alSuggested.add("org.thoughtcrime.securesms");
        alSuggested.add("com.whatsapp");
        alSuggested.add("com.instagram.android");
        alSuggested.add("im.vector.app");
        alSuggested.add("org.telegram.messenger");
        alSuggested.add("com.twitter.android");
        alSuggested.add("com.facebook.orca");
        alSuggested.add("com.facebook.mlite");
        alSuggested.add("com.brave.browser");
        alSuggested.add("org.mozilla.focus");

    }

    @Override
    protected void onResume() {
        super.onResume();
        mPrefs = Prefs.getSharedPrefs(getApplicationContext());
        reloadApps();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.app_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        if (item.getItemId() == R.id.menu_save_apps) {
            saveAppSettings();
            finish();
        }
        else if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void reloadApps() {
        new ReloadAppsAsyncTask(this).execute();
    }

    List<TorifiedApp> allApps, suggestedApps;
    List<TorifiedAppWrapper> uiList = new ArrayList<>();

    private void loadApps() {
        if (allApps == null)
            allApps = getApps(AppManagerActivity.this, mPrefs, null, alSuggested);

        TorifiedApp.sortAppsForTorifiedAndAbc(allApps);

        if (suggestedApps == null)
            suggestedApps = getApps(AppManagerActivity.this, mPrefs, alSuggested, null);

        final LayoutInflater inflater = getLayoutInflater();
        // only show suggested apps, text, etc and other apps header if there are any suggested apps installed...
        if (suggestedApps.size() > 0) {
            TorifiedAppWrapper headerSuggested = new TorifiedAppWrapper();
            headerSuggested.header = getString(R.string.apps_suggested_title);
            uiList.add(headerSuggested);
            TorifiedAppWrapper subheaderSuggested = new TorifiedAppWrapper();
            subheaderSuggested.subheader = getString(R.string.app_suggested_subtitle);
            uiList.add(subheaderSuggested);
            for (TorifiedApp app : suggestedApps) {
                TorifiedAppWrapper taw = new TorifiedAppWrapper();
                taw.app = app;
                uiList.add(taw);
            }
            TorifiedAppWrapper headerAllApps = new TorifiedAppWrapper();
            headerAllApps.header = getString(R.string.apps_other_apps);
            uiList.add(headerAllApps);
        }
        for (TorifiedApp app : allApps) {
            TorifiedAppWrapper taw = new TorifiedAppWrapper();
            taw.app = app;
            uiList.add(taw);
        }

        adapterAppsAll = new ArrayAdapter<>(this, R.layout.layout_apps_item, R.id.itemtext, uiList) {

            @Override
            public View getView(int position, View convertView, ViewGroup parent) {

                ListEntry entry = null;

                if (convertView == null)
                    convertView = inflater.inflate(R.layout.layout_apps_item, parent, false);
                else
                    entry = (ListEntry) convertView.getTag();

                if (entry == null) {
                    // Inflate a new view
                    entry = new ListEntry();
                    entry.container = convertView.findViewById(R.id.appContainer);
                    entry.icon = convertView.findViewById(R.id.itemicon);
                    entry.box = convertView.findViewById(R.id.itemcheck);
                    entry.text = convertView.findViewById(R.id.itemtext);
                    entry.header = convertView.findViewById(R.id.tvHeader);
                    entry.subheader = convertView.findViewById(R.id.tvSubheader);
                    convertView.setTag(entry);
                }

                final TorifiedAppWrapper taw = uiList.get(position);

                if (taw.header != null) {
                    entry.header.setText(taw.header);
                    entry.header.setVisibility(View.VISIBLE);
                    entry.subheader.setVisibility(View.GONE);
                    entry.container.setVisibility(View.GONE);
                } else if (taw.subheader != null) {
                    entry.subheader.setVisibility(View.VISIBLE);
                    entry.subheader.setText(taw.subheader);
                    entry.container.setVisibility(View.GONE);
                    entry.header.setVisibility(View.GONE);
                } else {
                    TorifiedApp app = taw.app;
                    entry.header.setVisibility(View.GONE);
                    entry.subheader.setVisibility(View.GONE);
                    entry.container.setVisibility(View.VISIBLE);
                    if (entry.icon != null) {

                        try {
                            entry.icon.setImageDrawable(pMgr.getApplicationIcon(app.getPackageName()));
                            entry.icon.setTag(entry.box);
                            entry.icon.setOnClickListener(AppManagerActivity.this);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }

                    if (entry.text != null) {
                        entry.text.setText(app.getName());
                        entry.text.setTag(entry.box);
                        entry.text.setOnClickListener(AppManagerActivity.this);
                    }

                    if (entry.box != null) {
                        entry.box.setChecked(app.isTorified());
                        entry.box.setTag(app);
                        entry.box.setOnClickListener(AppManagerActivity.this);

                    }
                }

                convertView.setOnFocusChangeListener((v, hasFocus) -> {
                    if (hasFocus)
                        v.setBackgroundColor(getResources().getColor(R.color.dark_purple));
                    else
                    {
                        v.setBackgroundColor(getResources().getColor(android.R.color.transparent));
                    }
                });

                return convertView;
            }
        };
    }

    public static ArrayList<TorifiedApp> getApps(Context context, SharedPreferences prefs, ArrayList<String> filterInclude, ArrayList<String> filterRemove) {

        PackageManager pMgr = context.getPackageManager();
        String tordAppString = prefs.getString(PREFS_KEY_TORIFIED, "");

        String[] tordApps;

        StringTokenizer st = new StringTokenizer(tordAppString, "|");
        tordApps = new String[st.countTokens()];
        int tordIdx = 0;
        while (st.hasMoreTokens()) {
            tordApps[tordIdx++] = st.nextToken();
        }
        Arrays.sort(tordApps);
        List<ApplicationInfo> lAppInfo = pMgr.getInstalledApplications(0);
        Iterator<ApplicationInfo> itAppInfo = lAppInfo.iterator();
        ArrayList<TorifiedApp> apps = new ArrayList<>();

        while (itAppInfo.hasNext()) {
            ApplicationInfo aInfo = itAppInfo.next();
            if (!includeAppInUi(aInfo)) continue;

            if (filterInclude != null) {
                boolean wasFound = false;
                for (String filterId : filterInclude)
                    if (filterId.equals(aInfo.packageName)) {
                        wasFound = true;
                        break;
                    }

                if (!wasFound)
                     continue;
            }

            if (filterRemove != null) {
                boolean wasFound = false;
                for (String filterId : filterRemove)
                    if (filterId.equals(aInfo.packageName)) {
                        wasFound = true;
                        break;
                    }

                if (wasFound)
                    continue;
            }

            TorifiedApp app = new TorifiedApp();

            try {
                PackageInfo pInfo = pMgr.getPackageInfo(aInfo.packageName, PackageManager.GET_PERMISSIONS);
                if (pInfo != null && pInfo.requestedPermissions != null) {
                    for (String permInfo : pInfo.requestedPermissions) {
                        if (permInfo.equals(Manifest.permission.INTERNET)) {
                            app.setUsesInternet(true);
                        }
                    }
                }


            } catch (Exception e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }

            try {
                app.setName(pMgr.getApplicationLabel(aInfo).toString());
            } catch (Exception e) {
                // no name
                continue; //we only show apps with names
            }


            if (!app.usesInternet())
                continue;
            else {
                apps.add(app);
            }

            app.setEnabled(aInfo.enabled);
            app.setUid(aInfo.uid);
            app.setUsername(pMgr.getNameForUid(app.getUid()));
            app.setProcname(aInfo.processName);
            app.setPackageName(aInfo.packageName);


            // check if this application is allowed
            app.setTorified(Arrays.binarySearch(tordApps, app.getPackageName()) >= 0);

        }

        Collections.sort(apps);

        return apps;
    }

    private void saveAppSettings() {

        StringBuilder tordApps = new StringBuilder();
        Intent response = new Intent();

        for (TorifiedApp tApp : allApps) {
            if (tApp.isTorified()) {
                tordApps.append(tApp.getPackageName());
                tordApps.append("|");
                response.putExtra(tApp.getPackageName(), true);
            }
        }

        for (TorifiedApp tApp : suggestedApps) {
            if (tApp.isTorified()) {
                tordApps.append(tApp.getPackageName());
                tordApps.append("|");
                response.putExtra(tApp.getPackageName(), true);
            }
        }

        Editor edit = mPrefs.edit();
        edit.putString(PREFS_KEY_TORIFIED, tordApps.toString());
        edit.apply();

        setResult(RESULT_OK, response);
    }

    public void onClick(View v) {

        CheckBox cbox = null;

        if (v instanceof CheckBox)
            cbox = (CheckBox) v;
        else if (v.getTag() instanceof CheckBox)
            cbox = (CheckBox) v.getTag();
        else if (v.getTag() instanceof ListEntry)
            cbox = ((ListEntry) v.getTag()).box;

        if (cbox != null) {
            final TorifiedApp app = (TorifiedApp) cbox.getTag();
            if (app != null) {
                app.setTorified(!app.isTorified());
                cbox.setChecked(app.isTorified());
            }

        }
    }

    private static class ReloadAppsAsyncTask extends AsyncTask<Void, Void, Void> {

        private final WeakReference<AppManagerActivity> activity;

        ReloadAppsAsyncTask(AppManagerActivity activity) {
            this.activity = new WeakReference<>(activity);
        }

        @Override
        protected void onPreExecute() {
            if (shouldStop()) return;
            activity.get().progressBar.setVisibility(View.VISIBLE);
        }

        @Override
        protected Void doInBackground(Void... voids) {
            if (shouldStop()) return null;
            activity.get().loadApps();
            return null;
        }

        @Override
        protected void onPostExecute(Void unused) {
            if (shouldStop()) return;
            AppManagerActivity ama = activity.get();
            ama.listAppsAll.setAdapter(ama.adapterAppsAll);
//            ama.listAppsAll.setAdapter(ama.adapterAppsAll);
            ama.progressBar.setVisibility(View.GONE);
        }

        private boolean shouldStop() {
            AppManagerActivity ama = activity.get();
            return ama == null || ama.isFinishing();
        }

    }

    private static class ListEntry {
        private CheckBox box;
        private TextView text; // app name
        private ImageView icon;

        private View container;
        private TextView header, subheader;
    }

}

File: app/src/main/java/org/torproject/android/ui/AppManagerActivity.kt
/* Copyright (c) 2009, Nathan Freitas, Orbot / The Guardian Project - http://openideals.com/guardian */ /* See LICENSE for licensing information */
package org.torproject.android.ui

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView

import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

import org.torproject.android.BuildConfig
import org.torproject.android.R
import org.torproject.android.service.OrbotConstants
import org.torproject.android.service.util.Prefs
import org.torproject.android.service.vpn.TorifiedApp

class AppManagerActivity : AppCompatActivity(), View.OnClickListener, OrbotConstants {
    inner class TorifiedAppWrapper {
        var header: String? = null
        var subheader: String? = null
        var app: TorifiedApp? = null
    }

    private var pMgr: PackageManager? = null
    private var mPrefs: SharedPreferences? = null
    private var recyclerView: RecyclerView? = null
    private var adapterAppsAll: RecyclerView.Adapter<*>? = null
    private var progressBar: ProgressBar? = null
    private var alSuggested: List<String>? = null
    private var searchBar: TextInputEditText? = null

    private var allApps: List<TorifiedApp>? = null
    private var suggestedApps: List<TorifiedApp>? = null
    private var uiList: MutableList<TorifiedAppWrapper> = ArrayList()

    private val job = Job()
    private val scope = CoroutineScope(Dispatchers.Main + job)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        pMgr = packageManager
        this.setContentView(R.layout.layout_apps)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        recyclerView = findViewById(R.id.applistview)
        progressBar = findViewById(R.id.progressBar)
        searchBar = findViewById(R.id.searchBar)

        // Need a better way to manage this list
        alSuggested = OrbotConstants.VPN_SUGGESTED_APPS

        searchBar?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                filterApps(s.toString())
            }
            override fun afterTextChanged(s: Editable?) {}
        })

        val searchBarLayout = findViewById<TextInputLayout>(R.id.searchBarLayout)
        searchBarLayout.setEndIconOnClickListener {
            searchBar?.text?.clear()
            searchBar?.clearFocus()
            hideKeyboard()
            reloadApps()
        }

        recyclerView?.layoutManager = LinearLayoutManager(this)
    }

    override fun onResume() {
        super.onResume()
        mPrefs = Prefs.getSharedPrefs(applicationContext)
        reloadApps()
    }

    override fun onClick(v: View) {
        val cbox = when (v) {
            is CheckBox -> v
            else -> v.tag as? CheckBox
        }

        cbox?.let {
            val app = it.tag as TorifiedApp
            app.isTorified = !app.isTorified
            it.isChecked = app.isTorified
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        val inflater = menuInflater
        inflater.inflate(R.menu.app_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_save_apps) {
            saveAppSettings()
            finish()
        } else if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
        job.cancel()
    }

    private fun hideKeyboard() {
        val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(searchBar?.windowToken, 0)
    }

    private fun calculateAppListHash(apps: List<TorifiedApp>?): Int {
        return apps?.sumOf { it.packageName.hashCode() } ?: -1
    }

    private fun reloadApps() {
        val sharedPreferences = getSharedPreferences("app_prefs", Context.MODE_PRIVATE)
        val cachedAppListHash = sharedPreferences.getString("cachedAppListHash", null)?.toIntOrNull()

        scope.launch {
            val currentAppListHash = calculateAppListHash(getApps(mPrefs, null, alSuggested))

            if (currentAppListHash != cachedAppListHash) {
                progressBar?.visibility = View.VISIBLE
                loadAppsAsync(fromCache = false)
                sharedPreferences.edit().putString("cachedAppListHash", currentAppListHash.toString()).apply()
                progressBar?.visibility = View.GONE
            } else {
                loadAppsAsync(fromCache = true)
            }
            recyclerView?.adapter = adapterAppsAll
        }
    }

    private suspend fun loadAppsAsync(fromCache: Boolean) {
        withContext(Dispatchers.Default) {
            if (fromCache) {
                val sharedPreferences = getSharedPreferences("app_prefs", Context.MODE_PRIVATE)
                val allAppsJson = sharedPreferences.getString("allApps", null)
                val suggestedAppsJson = sharedPreferences.getString("suggestedApps", null)
                allApps = Gson().fromJson(allAppsJson, object : TypeToken<List<TorifiedApp>>() {}.type)
                suggestedApps = Gson().fromJson(suggestedAppsJson, object : TypeToken<List<TorifiedApp>>() {}.type)
            } else {
                allApps = getApps(mPrefs, null, alSuggested)
                TorifiedApp.sortAppsForTorifiedAndAbc(allApps)
                suggestedApps = getApps(mPrefs, alSuggested, null)
                saveAppsToPrefs(allApps, suggestedApps)
            }

            val tordApps = mPrefs?.getString(OrbotConstants.PREFS_KEY_TORIFIED, "")?.split("|")?.sorted() ?: emptyList()
            allApps?.forEach { it.isTorified = tordApps.contains(it.packageName) }
            suggestedApps?.forEach { it.isTorified = tordApps.contains(it.packageName) }

            populateUiList()
            adapterAppsAll = createAdapter(uiList)
        }
    }

    private fun saveAppsToPrefs(allApps: List<TorifiedApp>?, suggestedApps: List<TorifiedApp>?) {
        val sharedPreferences = getSharedPreferences("app_prefs", Context.MODE_PRIVATE)
        val editor = sharedPreferences.edit()
        editor.putString("allApps", Gson().toJson(allApps))
        editor.putString("suggestedApps", Gson().toJson(suggestedApps))
        editor.apply()
    }

    private fun populateUiList() {
        uiList.clear()
        if (!suggestedApps.isNullOrEmpty()) {
            val headerSuggested = TorifiedAppWrapper()
            headerSuggested.header = getString(R.string.apps_suggested_title)
            uiList.add(headerSuggested)

            val subheaderSuggested = TorifiedAppWrapper()
            subheaderSuggested.subheader = getString(R.string.app_suggested_subtitle)
            uiList.add(subheaderSuggested)

            suggestedApps?.mapTo(uiList) { TorifiedAppWrapper().apply { app = it } }

            val headerAllApps = TorifiedAppWrapper()
            headerAllApps.header = getString(R.string.apps_other_apps)
            uiList.add(headerAllApps)
        }
        allApps?.mapTo(uiList) { TorifiedAppWrapper().apply { app = it } }
    }

    class AppViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val container: ViewGroup = view.findViewById(R.id.appContainer)
        val icon: ImageView = view.findViewById(R.id.itemicon)
        val box: CheckBox = view.findViewById(R.id.itemcheck)
        val text: TextView = view.findViewById(R.id.itemtext)
        val header: TextView = view.findViewById(R.id.tvHeader)
        val subheader: TextView = view.findViewById(R.id.tvSubheader)
    }

    class AppAdapter(
        private val list: List<TorifiedAppWrapper>,
        private val pMgr: PackageManager,
        private val onClickListener: View.OnClickListener
    ) : RecyclerView.Adapter<AppViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AppViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.layout_apps_item, parent, false)
            return AppViewHolder(view)
        }

        override fun onBindViewHolder(holder: AppViewHolder, position: Int) {
            val taw = list[position]
            if (taw.header != null) {
                holder.header.text = taw.header
                holder.header.visibility = View.VISIBLE
                holder.subheader.visibility = View.GONE
                holder.container.visibility = View.GONE
            } else if (taw.subheader != null) {
                holder.subheader.visibility = View.VISIBLE
                holder.subheader.text = taw.subheader
                holder.container.visibility = View.GONE
                holder.header.visibility = View.GONE
            } else {
                val app = taw.app
                holder.header.visibility = View.GONE
                holder.subheader.visibility = View.GONE
                holder.container.visibility = View.VISIBLE
                try {
                    holder.icon.setImageDrawable(pMgr.getApplicationIcon(app!!.packageName))
                    holder.icon.tag = holder.box
                    holder.icon.setOnClickListener(onClickListener)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                holder.text.text = app!!.name
                holder.text.tag = holder.box
                holder.text.setOnClickListener(onClickListener)
                holder.box.isChecked = app.isTorified
                holder.box.tag = app
                holder.box.setOnClickListener(onClickListener)
            }
        }

        override fun getItemCount(): Int = list.size
    }

    private fun createAdapter(list: List<TorifiedAppWrapper>): RecyclerView.Adapter<AppViewHolder> {
        return AppAdapter(list, pMgr!!, this)
    }

    private fun filterApps(query: String) {
        val filteredList = uiList.filter {
            it.app?.name?.contains(query, ignoreCase = true) == true
        }
        adapterAppsAll = createAdapter(filteredList)
        recyclerView?.adapter = adapterAppsAll
    }

    private fun saveAppSettings() {
        val tordApps = StringBuilder()
        val response = Intent()

        val saveTorifiedApps: (List<TorifiedApp>?) -> Unit = { apps ->
            apps?.filter { it.isTorified }?.forEach { tApp ->
                tordApps.append(tApp.packageName).append("|")
                response.putExtra(tApp.packageName, true)
            }
        }

        saveTorifiedApps(allApps)
        saveTorifiedApps(suggestedApps)

        mPrefs?.edit()?.apply {
            putString(OrbotConstants.PREFS_KEY_TORIFIED, tordApps.toString().trimEnd('|'))
            apply()
        }

        setResult(RESULT_OK, response)
    }

    private fun getApps(
        prefs: SharedPreferences?,
        filterInclude: List<String>?,
        filterRemove: List<String>?
    ): ArrayList<TorifiedApp> {
        val tordApps = prefs?.getString(OrbotConstants.PREFS_KEY_TORIFIED, "")?.split("|")?.sorted() ?: emptyList()
        val apps = ArrayList<TorifiedApp>()

        pMgr?.getInstalledApplications(0)?.forEach { aInfo ->
            if (!aInfo.enabled ||
                OrbotConstants.BYPASS_VPN_PACKAGES.contains(aInfo.packageName) ||
                BuildConfig.APPLICATION_ID == aInfo.packageName) return@forEach

            if (filterInclude != null && aInfo.packageName !in filterInclude) return@forEach
            if (filterRemove != null && aInfo.packageName in filterRemove) return@forEach

            val app = TorifiedApp().apply {
                try {
                    pMgr?.getPackageInfo(aInfo.packageName, PackageManager.GET_PERMISSIONS)?.requestedPermissions?.let { perms ->
                        if (Manifest.permission.INTERNET in perms) setUsesInternet(true)
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }

                try {
                    name = pMgr?.getApplicationLabel(aInfo).toString()
                } catch (e: Exception) {
                    return@forEach
                }

                if (!usesInternet()) return@forEach

                isEnabled = aInfo.enabled
                uid = aInfo.uid
                username = pMgr?.getNameForUid(uid)
                procname = aInfo.processName
                packageName = aInfo.packageName
                isTorified = tordApps.contains(packageName)
            }

            apps.add(app)
        }

        apps.sort()
        return apps
    }
}
